kiosk/src/share.ts

```diff
@@ -0,0 +1,93 @@
+import { GameData } from "./Models/GameData";
+
+const apiRoot = "https://www.makecode.com";
+const description = "A kiosk for MakeCode Arcade";
+
+export interface SharedKioskData {
+    games: GameData[];
+}
+
+export async function createKioskShareLink(kioskData: SharedKioskData) {
+    const payload = createShareRequest(
+        "kiosk",
+        createProjectFiles("kiosk", kioskData)
+    );
+    const url = apiRoot + "/api/scripts";
+
+    const result = await fetch(
+        url,
+        {
+            method: "POST",
+            body: new Blob([JSON.stringify(payload)], { type: "application/json" })
+        }
+    );
+
+    if (result.status === 200) {
+        const resJSON = await result.json();
+        // return "https://arcade.makecode.com/" + resJSON.shortid;
+        return `${resJSON.shortid}/kiosk.json`;
+    }
+
+    return "ERROR"
+}
+
+
+function createShareRequest(projectName: string, files: {[index: string]: string}) {
+    const header = {
+        "name": projectName,
+        "meta": {
+        },
+        "editor": "tsprj",
+        "pubId": undefined,
+        "pubCurrent": false,
+        "target": "arcade",
+        "id": crypto.randomUUID(),
+        "recentUse": Date.now(),
+        "modificationTime": Date.now(),
+        "path": projectName,
+        "saveId": {},
+        "githubCurrent": false,
+        "pubVersions": []
+    }
+
+    return {
+        id: header.id,
+        name: projectName,
+        target: header.target,
+        description: description,
+        editor: "tsprj",
+        header: JSON.stringify(header),
+        text: JSON.stringify(files),
+        meta: {
+        }
+    }
+}
+
+function createProjectFiles(projectName: string, kioskData: SharedKioskData) {
+    const files: {[index: string]: string} = {};
+
+    const config = {
+        "name": projectName,
+        "description": description,
+        "dependencies": {
+            "device": "*"
+        },
+        "files": [
+            "main.ts",
+            "kiosk.json"
+        ],
+        "preferredEditor": "tsprj"
+    };
+    files["pxt.json"] = JSON.stringify(config, null, 4);
+    files["main.ts"] = " ";
+    files["kiosk.json"] = JSON.stringify(kioskData, undefined, 4);
```

Comment:
Asking for more understanding. Why are each of these files needed? Are these the files a share link is expected to serve so to speak?

Reply:
`kiosk.json` and `pxt.json` are actually necessary; `pxt.json` is the project config and is required to be a valid makecode project (I'm guessing the post request would get rejected if we don't have a pxt.json in the blob but haven't tried; could just look at the backend code to confirm). `kiosk.json` here is where we're storing the data we want to share (kiosk state), so that's important. I left an empty `main.ts` just so there wouldn't be any weirdness if someone chose to open in the editor, not sure the behavior we would get from opening a shared project with no blocks/ts/py files in it off the top of my head and easiest just to include main.ts that is in almost every project rather than think about it (easy enough to remove from `files` & `config.files` to check though)